rfqservice/negotiator.go

```diff
@@ -0,0 +1,279 @@
+		)
+		var msg rfqmsg.OutgoingMsg = &rejectMsg
+
+		sendSuccess := fn.SendOrQuit(n.outgoingMessages, msg, n.Quit)
+		if !sendSuccess {
+			return fmt.Errorf("negotiator failed to send reject " +
+				"message")
+		}
+	}
+
+	// Query the price oracle in a separate goroutine in case it is a remote
+	// service and takes a long time to respond.
+	n.Wg.Add(1)
+	go func() {
+		defer n.Wg.Done()
+
+		oracleResp, err := n.queryPriceOracle(req)
+		if err != nil {
+			log.Errorf("negotiator failed to query price oracle: "+
+				"%v", err)
+		}
+
+		err = n.handlePriceOracleResponse(req, oracleResp)
+		if err != nil {
+			log.Errorf("negotiator failed to handle price oracle "+
+				"response: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+// AssetSellOffer is a struct that represents an asset sell offer. This
+// data structure describes the maximum amount of an asset that is available
+// for sale.
+type AssetSellOffer struct {
+	// AssetID represents the identifier of the subject asset.
+	AssetID *asset.ID
+
+	// AssetGroupKey is the public group key of the subject asset.
+	AssetGroupKey *btcec.PublicKey
+
+	// MaxAssetAmount is the maximum amount of the asset under offer.
+	MaxAssetAmount uint64
+}
+
+// Validate validates the asset sell offer.
+func (a *AssetSellOffer) Validate() error {
+	if a.AssetID == nil && a.AssetGroupKey == nil {
+		return fmt.Errorf("asset ID is nil and asset group key is nil")
+	}
+
+	if a.AssetID != nil && a.AssetGroupKey != nil {
+		return fmt.Errorf("asset ID and asset group key are both set")
+	}
+
+	if a.MaxAssetAmount == 0 {
+		return fmt.Errorf("max asset amount is zero")
+	}
+
+	return nil
+}
+
+// UpsertAssetSellOffer upserts an asset sell offer with the negotiator.
+func (n *Negotiator) UpsertAssetSellOffer(offer AssetSellOffer) error {
+	// Validate the offer.
+	err := offer.Validate()
+	if err != nil {
+		return fmt.Errorf("invalid asset sell offer: %w", err)
+	}
+
+	// Store the offer in the appropriate map.
+	//
+	// If the asset group key is not nil, then we will use it as the key for
+	// the offer. Otherwise, we will use the asset ID as the key.
+	switch {
+	case offer.AssetGroupKey != nil:
+		compressedKey := offer.AssetGroupKey.SerializeCompressed()
+		keyStr := hex.EncodeToString(compressedKey)
+
+		n.assetGroupSellOffers[keyStr] = offer
+
+	case offer.AssetID != nil:
+		idStr := offer.AssetID.String()
+		n.assetSellOffers[idStr] = offer
+	}
+
+	return nil
+}
+
+// RemoveAssetSellOffer removes an asset sell offer from the negotiator.
+func (n *Negotiator) RemoveAssetSellOffer(assetID *asset.ID,
+	assetGroupKey *btcec.PublicKey) error {
+
+	// Remove the offer from the appropriate map.
+	//
+	// If the asset group key is not nil, then we will use it as the key for
+	// the offer. Otherwise, we will use the asset ID as the key.
+	switch {
+	case assetGroupKey != nil:
+		compressedKey := assetGroupKey.SerializeCompressed()
+		keyStr := hex.EncodeToString(compressedKey)
+
+		delete(n.assetGroupSellOffers, keyStr)
```

Comment:
What about auto expiry?

Reply:
I'm not sure what we want here ultimately. I think that the latest commits have the simplest case for now.

I think we need to link this to the asset wallet somehow.